google-cloud-clients/google-cloud-spanner/src/test/java/com/google/cloud/spanner/TransactionRunnerImplTest.java

```diff
@@ -141,6 +154,82 @@ public void runResourceExhaustedNoRetry() throws Exception {
     verify(txn).rollback();
   }
 
+  @Test
+  public void batchDmlAborted() {
+    batchDmlException(Code.ABORTED_VALUE);
+  }
+
+  @Test
+  public void batchDmlFailedPrecondition() {
+    batchDmlException(Code.FAILED_PRECONDITION_VALUE);
+  }
+
+  @SuppressWarnings("unchecked")
+  private void batchDmlException(int status) {
+    Preconditions.checkArgument(status != Code.OK_VALUE);
+    TransactionContextImpl transaction =
+        new TransactionContextImpl(session, ByteString.copyFromUtf8("test"), rpc, 10);
+    when(session.newTransaction()).thenReturn(transaction);
+    when(session.getName()).thenReturn("test");
+    TransactionRunnerImpl runner = new TransactionRunnerImpl(session, rpc, 10);
+    ExecuteBatchDmlResponse response1 =
+        ExecuteBatchDmlResponse.newBuilder()
+            .addResultSets(
+                ResultSet.newBuilder()
+                    .setStats(ResultSetStats.newBuilder().setRowCountExact(1L))
+                    .build())
+            .setStatus(com.google.rpc.Status.newBuilder().setCode(status).build())
+            .build();
+    ExecuteBatchDmlResponse response2 =
+        ExecuteBatchDmlResponse.newBuilder()
+            .addResultSets(
+                ResultSet.newBuilder()
+                    .setStats(ResultSetStats.newBuilder().setRowCountExact(1L))
+                    .build())
+            .addResultSets(
+                ResultSet.newBuilder()
+                    .setStats(ResultSetStats.newBuilder().setRowCountExact(1L))
+                    .build())
+            .setStatus(com.google.rpc.Status.newBuilder().setCode(Code.OK_VALUE).build())
+            .build();
+    when(rpc.executeBatchDml(Mockito.any(ExecuteBatchDmlRequest.class), Mockito.anyMap()))
+        .thenReturn(response1, response2);
+    CommitResponse commitResponse =
+        CommitResponse.newBuilder().setCommitTimestamp(Timestamp.getDefaultInstance()).build();
+    when(rpc.commit(Mockito.any(CommitRequest.class), Mockito.anyMap())).thenReturn(commitResponse);
+    final Statement statement = Statement.of("UPDATE FOO SET BAR=1");
+    final AtomicInteger numCalls = new AtomicInteger(0);
+    SpannerBatchUpdateException exception = null;
+    long updateCount[];
+    try {
+      updateCount =
+          runner.run(
+              new TransactionCallable<long[]>() {
+                @Override
+                public long[] run(TransactionContext transaction) throws Exception {
+                  numCalls.incrementAndGet();
+                  return transaction.batchUpdate(Arrays.asList(statement, statement));
+                }
+              });
+    } catch (SpannerBatchUpdateException e) {
+      exception = e;
+      updateCount = e.getUpdateCounts();
+    }
+    if (status == Code.ABORTED_VALUE) {
```

Comment:
Is there anyway to move the asserts into `batchDmlAborted` and `batchDmlFailedPrecondition`?